Scroll viewport when block on edge

bcjordan opened the pull request:
# Scrolls viewport when block is on edge

![scroll-based-on-block](https://cloud.githubusercontent.com/assets/206973/8918701/97ea6ae8-346e-11e5-9b06-1287e9aba673.gif)

This adds a number of parameters exposed in `scroll_on_block_drag_handler.js`:

MOUSE_SPEED_SLOW/FAST: speeds in pixels/sec to move when mouse is just beyond (slow) or way beyond (fast) the edge.

BLOCK_SPEED_SLOW/FAST: speeds in pixels/sec to move when block's boundaries are just beyond (slow) or way beyond (fast) the edge.
# Scrolls block into view when dropped beyond viewport bounds

![pantoseeblock](https://cloud.githubusercontent.com/assets/206973/8917283/fa638528-3465-11e5-8ac7-3c09e5fdb357.gif)
# Fix scrolling behavior when a boundary-defining block is picked up and dragged around

Previously, when a block was scroll-boundary-defining (i.e., one of the rightmost or bottommost), when the block was picked up and the blockspace was scrolled, the scrolling viewport would auto-resize and often jump to the top of the screen.

This includes a fix to no longer require.
# Add ability to bump entire blocks back into view

Our block bumping logic currently bumps blocks dropped off the right and bottom sides of the screen just far enough to be grabbable. There is now a flag (and debug test checkbox in the playground) to enable a different behavior which bumps the entire block into view.

![bumpentireblock](https://cloud.githubusercontent.com/assets/206973/8917308/233fffda-3466-11e5-976d-f87c8b12dd00.gif)

Since this new behavior might cramp workspaces when there is no scrolling enabled, it may make sense to turn it on after some feel-testing and as scrolling is enabled.

islemaster commented on blockly-core/core/ui/block.js:
```diff
@@ -1077,12 +1093,12 @@ Blockly.Block.prototype.onMouseMove_ = function(e) {
     var y = this.startDragY + dy;
     this.svg_.getRootElement().setAttribute('transform',
-        'translate(' + x + ', ' + y + ')');
+      'translate(' + x + ', ' + y + ')');
```
Did your editor settings update to two-space line continuation indent or something?

islemaster commented on blockly-core/core/ui/block_space/auto_scroll.js:
```diff
@@ -0,0 +1,105 @@
+
+/**
+ * @fileoverview Represents an active, update-able automatic scroll behavior.
+ * @author fraser@google.com (Neil Fraser)
```
Doesn't look like you pulled this from upstream, might want to update the authorship metadata.

islemaster commented on blockly-core/core/ui/block_space/auto_scroll.js:
```diff
@@ -0,0 +1,105 @@
+
+Blockly.AutoScroll.prototype.scrollTick_ = function (msPassed) {
+  var xPixelsPerMs = (this.activePanVector_.x / 1000);
+  var yPixelsPerMs = (this.activePanVector_.y / 1000);
```
Nit: It would be nice if these were precalculated whenever you set `this.activePanVector_` instead of doing this division on every tick.

islemaster commented on blockly-core/core/ui/block_space/block_space.js:
```diff
@@ -819,11 +858,66 @@ Blockly.BlockSpace.prototype.getScrollableSize = function(metrics) {
+    boxOverflows.bottom - boxOverflows.top);
+};
+
+Blockly.BlockSpace.prototype.scrollDeltaWithAnySelectedBlock = function (scrollDx, scrollDy,
```
Needs doc - I'm not sure from the name what the mouseX and mouseY parameters do.

islemaster replied:
Just saw the nice comment on scrollWithAnySelectedBlock below.  Might be sufficient to refer to that doc here.

islemaster commented:
Making progress, but headed to meeting.  Want to pull this locally and try it out before I approve.

islemaster commented:
I get this exception with "Toggle bump debug drawing enabled" on the playground.  Not sure if I set something up wrong.

![screenshot from 2015-07-27 19 20 24](https://cloud.githubusercontent.com/assets/1615761/8922186/9d5355dc-3494-11e5-8609-6d0e4f4e8696.png)

islemaster commented:
When disabling "Toggle scroll-on-drag debug drawing" in the playground, the debug rects don't go away, they just stop updating.

islemaster commented on blockly-core/core/ui/block_space/scroll_on_block_drag_handler.js:
```diff
@@ -0,0 +1,215 @@
+goog.require('goog.math.Vec2');
+
+/**
+ * @param {!Blockly.ScrollOnBlockDragHandler} blockSpace
```
This type looks incorrect.

islemaster commented on blockly-core/core/ui/block_space/scroll_on_block_drag_handler.js:
```diff
@@ -0,0 +1,215 @@
+  /**
+   * BlockSpace to scroll
+   * @type {Blockly.BlockSpace}
+   * @private
```
Phil pointed out to me the other day that `@private {Blockly.BlockSpace}` is a valid, more compact version of this (when using closure compiler).

islemaster commented:
Okay @bcjordan I'm done with my review pass.  This looks great, and feels great (tried it locally, very smooth).  I did notice some issues with the debug drawing stuff in the playground - obviously noncritical, but we want to keep that in good shape since it's been so useful.

bcjordan commented:
@islemaster thanks! Implemented suggested changes and fixed up the debug drawing behavior a bit.

Debug button and toggle debug-drawing on bump type:

![debugdraw](https://cloud.githubusercontent.com/assets/206973/8967562/7f6e5a9a-35ec-11e5-8909-dc107780fe64.gif)

Can you take another look?

bcjordan commented:
(Merging into `vertical_scrollbars` branch pre-final review)

islemaster commented:
Looks great! :+1:
